app/src/main/java/io/apicurio/registry/storage/impl/jdbc/JdbcRegistryStorage.java

```diff
@@ -0,0 +1,646 @@
+                .getSingleResult() != 0;
+    }
+
+    private boolean _artifactVersionExists(String artifactId, long version) {
+        return entityManager.createQuery("SELECT COUNT(a) FROM Artifact a " +
+                "WHERE a.artifactId = :artifact_id AND a.version = :version", Long.class)
+                .setParameter("artifact_id", artifactId)
+                .setParameter("version", version)
+                .getSingleResult() != 0;
+    }
+
+
+    // ========================================================================
+
+
+    @Override
+    @Transactional
+    public ArtifactMetaDataDto createArtifact(String artifactId, ArtifactType artifactType, String content) throws ArtifactAlreadyExistsException, RegistryStorageException {
+        requireNonNull(artifactId);
+        requireNonNull(artifactType);
+        requireNonNull(content);
+
+        try {
+            long nextVersion = _getNextArtifactVersion(artifactId);
+
+            if (nextVersion != 1) {
+                throw new ArtifactAlreadyExistsException(artifactId);
+            }
+
+            Artifact artifact = Artifact.builder()
+                    .artifactId(artifactId)
+                    .version(nextVersion)
+                    .content(content)
+                    .build();
+
+            entityManager.persist(artifact);
+
+            new MetaDataMapperUpdater()
+                    .update(MetaDataKeys.TYPE, artifactType.value())
+                    .persistUpdate(entityManager, artifactId, nextVersion);
+
+            return new MetaDataMapperUpdater()
+                    .update(MetaDataKeys.TYPE, artifactType.value())
+                    .persistUpdate(entityManager, artifactId, null)
+                    .update(artifact)
+                    .toArtifactMetaDataDto();
+
+        } catch (PersistenceException ex) {
+            throw new RegistryStorageException(ex);
+        }
+    }
+
+    @Override
+    @Transactional
+    public SortedSet<Long> deleteArtifact(String artifactId) throws ArtifactNotFoundException, RegistryStorageException {
+        try {
+
+            if (!_artifactExists(artifactId))
+                throw new ArtifactNotFoundException(artifactId);
+
+            List<Long> res1 = entityManager.createQuery(
+                    "SELECT a.version FROM Artifact a " +
+                            "WHERE a.artifactId = :artifact_id " +
+                            "ORDER BY a.version DESC", Long.class)
+                    .setParameter("artifact_id", artifactId)
+                    .getResultList();
+
+            if (res1.size() == 0)
+                throw new ArtifactNotFoundException(artifactId);
+
+            int affected = entityManager.createQuery(
+                    "DELETE FROM Artifact a " +
+                            "WHERE a.artifactId = :artifact_id")
+                    .setParameter("artifact_id", artifactId)
+                    .executeUpdate();
+
+            if (res1.size() != affected)
```

Comment:
I would just do a delete, and if that returns 0, you can then throw ArtifactNotFoundException.

Reply:
The method has to return the deleted versions, so I need to select them anyway. The assertion is probably not necessary though.